engine/chunker/cached_chunker.go

```diff
@@ -0,0 +1,516 @@
+	}
+	return nil
+}
+
+// performOnCache is a utility to perform operatons in CachedEntriesChunker.cache to safely set
+// the context to be used during eviction and return errors that may occur as a result of
+// eviction if performing the given action indeed causes it.
+func (ls *CachedEntriesChunker) performOnCache(ctx context.Context, action func(*lru.Cache)) error {
+	ls.onEvictedCtx = ctx
+	defer func() {
+		ls.onEvictedCtx = nil
+		ls.onEvictedErr = nil
+	}()
+	action(ls.cache)
+	err := ls.onEvictedErr
+	return err
+}
+
+// listEntriesChainLinks lists the links to the entries chain with given root.
+// The returned list will always include the root itself.
+//
+// Note that if traversal of the chain partially fails any links listed so far will be returned
+// along with the error.
+func (ls *CachedEntriesChunker) listEntriesChainLinks(ctx context.Context, root ipld.Link) ([]ipld.Link, error) {
+	var links []ipld.Link
+	lCtx := ipld.LinkContext{Ctx: ctx}
+	next := root
+	for {
+		n, err := ls.lsys.Load(lCtx, next, schema.Type.EntryChunk)
+		if err != nil {
+			return links, err
+		}
+		chunk, ok := n.(schema.EntryChunk)
+		if !ok {
+			return links, fmt.Errorf("expected schema.EntryChunk for link: %s", next)
+		}
+		links = append(links, next)
+
+		if chunk.FieldNext().IsAbsent() || chunk.FieldNext().IsNull() {
+			break
+		}
+		lnk, err := chunk.FieldNext().AsNode().AsLink()
+		if err != nil {
+			return links, err
+		}
+		next = lnk
+	}
+	return links, nil
+}
+
+// Cap returns the maximum number of chained entries chunks this cache stores.
+//
+// Note, the maximum number refers to the number of chains as a unit and not the total sum of
+// individual chunks across chains.
+func (ls *CachedEntriesChunker) Cap() int {
+	return ls.cache.MaxEntries
+}
+
+// Len returns the number of chained entries chunks thar are currently stored in cache.
+//
+// Note, the number refers to the number of chains as a unit and not the total sum of individual
+// chunks across chains.
+func (ls *CachedEntriesChunker) Len() int {
+	ls.lock.Lock()
+	defer ls.lock.Unlock()
+	return ls.cache.Len()
+}
+
+func (ls *CachedEntriesChunker) dsRootPrefixedKey(l ipld.Link) datastore.Key {
+	return rootKeyPrefix.Child(dsKey(l))
+}
+
+func (ls *CachedEntriesChunker) linkFromDsCachePrefixedKey(ck datastore.Key) (ipld.Link, error) {
+
+	if !rootKeyPrefix.IsAncestorOf(ck) {
+		return nil, fmt.Errorf("key is not a prefixed cache key: %s", ck)
+	}
+	c, err := cid.Decode(ck.BaseNamespace())
+	if err != nil {
+		return nil, err
+	}
+
+	return cidlink.Link{Cid: c}, nil
+}
+
+func (ls *CachedEntriesChunker) incrementOverlap(ctx context.Context, lnk ipld.Link) error {
+	oKey := ls.dsOverlapPrefixedKey(lnk)
+	oVal, err := ls.ds.Get(ctx, oKey)
+	var count uint64
+	if err == datastore.ErrNotFound {
+		count = 1
+		oVal = make([]byte, 8)
+	} else if err != nil {
+		return err
+	} else {
+		count = binary.LittleEndian.Uint64(oVal) + 1
+	}
```

Comment:
```suggestion
	if err != nil {
		if err == datastore.ErrNotFound {
			count = 1
			oVal = make([]byte, 8)
		} else {
			return err
		}
	} else {
		count = binary.LittleEndian.Uint64(oVal) + 1
	}
```

nit: eliminate a comparison when `err == nil`